www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,230 @@
+
+```c
+	if (is_rel)
+		LockRelationOid(objid, lockmode);
+	else
+		LockDatabaseObject(classid, objid, objsubid, lockmode);
+```
+
+If we need to lock a table (also called relation), we use `LockRelationOid`, the Postgres method behind `LOCK TABLE`. If we need to lock another object, we use `LockDatabaseObject` and directly lock the object we need.
+
+This is simple enough when acquiring locks locally, but how do we distribute this logic? We wrapped the object lock acquisition logic in the `nile_ddl` extension in a function and then exported it as a function:
+
+```c
+PGDLLEXPORT Datum nile_ddl_objectlock(PG_FUNCTION_ARGS);
+
+PG_FUNCTION_INFO_V1(nile_ddl_objectlock);
+```
+
+Then the transaction coordinator can send `SELECT nile_ddl.nile_ddl_objectlock(...)` to all the databases and acquire all the necessary locks for the DDL.
+
+When the transaction coordinator distributes the actual DDL, Postgres will try to re-acquire the locks. Because the DDL execution will happen in the same session in which we acquired the locks initially, and the locks are on the same objects and with the same level, we are guaranteed that the DDL processing will successfully acquire these locks. 
+
+So far we dealt with acquiring locks on existing tables and other objects. But there is one other case we need to handle before we are done. `CREATE TABLE` also requires locks - in order to avoid conflict between two sessions attempting to create the same table. How can we lock a table that does not exist yet? 
+
+Normally Postgres uses the data dictionary for this. It inserts a row for this new table into several tables in pg_catalog, the table name is added to unique indexes, and any attempt to insert another row with the same name into the same unique index will first block and then fail after the first session commits. Unfortunately, we can't use the same method. Row locks behave differently from table locks - they can't get re-acquired. Which means that if we manage to successfully acquire such lock, the actual DDL will then block on the same lock. 
+
+Another option would be to use an "advisory lock". These are locks not enforced by Postgres itself, but they can be acquired and used by application logic. The problem is that there is a fixed number of advisory locks per transaction and we want to leave all of them to our users, since their application logic may depend on having these locks available.
```

Comment:
nit: the advisory lock address space is fixed, it doesn't depend on transactions (there's a limit on how many locks you can acquire concurrently but that's the same limit for any kind of heavyweight lock)

Reply:
I may be jumping to conclusions here, but looking at this doc: https://www.postgresql.org/docs/current/explicit-locking.html#ADVISORY-LOCKS

I see: 

> Both advisory locks and regular locks are stored in a shared memory pool whose size is defined by the configuration variables [max_locks_per_transaction](https://www.postgresql.org/docs/current/runtime-config-locks.html#GUC-MAX-LOCKS-PER-TRANSACTION) and [max_connections](https://www.postgresql.org/docs/current/runtime-config-connection.html#GUC-MAX-CONNECTIONS). Care must be taken not to exhaust this memory or the server will be unable to grant any locks at all. This imposes an upper limit on the number of advisory locks grantable by the server, typically in the tens to hundreds of thousands depending on how the server is configured.

`max_locks_per_transaction` implies both per-transaction limit, and together with `max_connections` also a global limit. You are saying that this is a general limit on heavy locks. And that there is an additional limit on advisory locks that isn't documented?

Reply:
mmm... by "address space" you may be refering to the identifies set in `pg_advisory_lock(id)`, but this accepts `bigint`, which seems like a much higher limit than the max_locks_per_transaction.

Reply:
I'll remove the "per transaction" from my blog as a precaution, but I'd love to understand this limit better.

Reply:
yes there are 2 different limits: the number of locks you can hold at the same time, which is computed using the 2 GUCs you mentioned. and then there's how many different advisory locks that can exist.

advisory locks is just a bigint under the hood while the other locks have more bits to express what to lock. 


so if we relied on advisory locks for our own user locks, we would need to use almost all the address space of the advisory locks, and even doing so the risk of collision would be higher (since we don't have an oid to lock we have to lock the hash of the object name) than our current approach. one thing that's mentioned but maybe not very obvious is that for our custom locks the first field is the virtual database (more precisely a well known oid within that virtual database), which guarantees that if a collision happens it will be inside a single virtual database, so no neighbour problem. this part already uses half the advisory lock address space.

Reply:
yes, but leaving the door opened for maybe eventually allowing advisory locks for clients is a nice side effect :)